### ISSUE TITLE
array_slice should convert non-empty-array/list to array/list?

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#2246]
Also note that array_slice on arrays with integer keys convert those arrays to lists.

If the offset is 0 and the limit is positive, the non-empty part could be preserved.

I expect other return type providers could have similar issues, but haven't checked.

```php
<?php

/** @return false|string */
function test(array $gIds, int $n) {
    if (empty($gIds)) {
        return 'empty';
    }
    if (rand(0,1)) {
        // Replacing this with `$gIds = array $other` for a different array param would fix the false positive.
        $gIds = array_slice($gIds, $n);
    }
    // Observed: TypeDoesNotContainType - src/redundant.php:11:9 - Found a contradiction when evaluating $gIds and trying to reconcile type 'non-empty-array<array-key, mixed>' to empty
    // Expected: No warning - the array slice could be empty
    if (empty($gIds)) {
        return 'empty $n';
    }
    return json_encode($gIds);
}
```
